### PR TITLE
Migrate nosekatreport from python2 to python3

### PR DESCRIPTION
**Pull request by bngcebetsha**
Ported nosekatreport package to python3, modified aqf tests to acount for the changes made.

[MT-936](https://skaafrica.atlassian.net/browse/MT-936)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,35 +1,57 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
 # Copyright (c) 2017 National Research Foundation (South African Radio Astronomy Observatory)
 # BSD license - see LICENSE for details
-from setuptools import setup, find_packages
+from future import standard_library
```
Comment (tockards) [inline setup.py]: @bngcebetsha and @amakhaba 
You want to remove this line as it is causing the build to fail 
https://travis-ci.org/ska-sa/nosekatreport/builds/580658530?utm_source=github_status&utm_medium=notification

As you are using `future` before the installation.

Comment (amakhaba) [reply setup.py]: Removed the import statement from `setup.py` but does not fix the problem.

File: setup.py [hunk 2]
```diff
@@ -1,35 +1,56 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
 # Copyright (c) 2017 National Research Foundation (South African Radio Astronomy Observatory)
 # BSD license - see LICENSE for details
-from setuptools import setup, find_packages
+from setuptools import find_packages, setup
+
+standard_library.install_aliases()
```
Comment (tockards) [inline setup.py]: @bngcebetsha @amakhaba  this line needs to be removed also.

File: setup.py [hunk 3]
```diff
@@ -1,35 +1,56 @@
...
+    platforms=["OS Independent"],
+    keywords="meerkat kat ska",
+    python_requires=">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, <4",
+    setup_requires=["katversion"],
+    use_katversion=True,
+    install_requires=[
+        "elasticsearch>=5.4.0",
+        "future",
+        "futures; python_version<'3'",
+        "katconf",
```
Comment (tockards) [inline setup.py]: @bngcebetsha and @amakhaba This will fail as the `katconf` repo is private. 

try adding the following after the `install_requires` list

```
      dependency_links=[
          'git+https://github.com/ska-sa/katconf.git'
      ],
```
See https://python-packaging.readthedocs.io/en/latest/dependencies.html#packages-not-on-pypi

This will fail as it will require access to the library but just add it to see the failure on travis

Comment (tockards) [reply setup.py]: @bngcebetsha  and @amakhaba 
Actually, You can remove the `katconf` fromt he required packages as it seems like it is not used in nosekatreport.

**Comment by mmphego** [comment]
@bngcebetsha @amakhaba use this `travis.yml` to test PY2 and PY3.

https://raw.githubusercontent.com/ska-sa/nosekatreport/4a6d6d0ae545a49178bda3d1c12c78ca8e45f594/.travis.yml

File: nosekatreport/decorators.py [hunk 4]
```diff
@@ -1,3 +1,10 @@
+from __future__ import (absolute_import, division, print_function)
+
+from future import standard_library
+standard_library.install_aliases()
+
+from builtins import str
```
Comment (mmphego) [inline nosekatreport/decorators.py]: Future string import in the past broke a lot of stuff from my end, not sure if you need it.
Run tests without it and confirm if tests pass, if they do - please remove.

File: nosekatreport/decorators.py [hunk 5]
```diff
@@ -1,3 +1,10 @@
...
+
 # Copyright (c) 2017 National Research Foundation (South African Radio Astronomy Observatory)
 # BSD license - see LICENSE for details
```
Comment (mmphego) [inline nosekatreport/decorators.py]: Move licensing information to the top of the file.

File: nosekatreport/plugin.py [hunk 6]
```diff
@@ -1,12 +1,11 @@
-# Copyright (c) 2017 National Research Foundation (South African Radio Astronomy Observatory)
```
Comment (mmphego) [inline nosekatreport/plugin.py:1]: Why did you remove the license information?

File: nosekatreport/plugin.py [hunk 7]
```diff
@@ -16,6 +15,14 @@
 import tempfile
 import time
 import traceback
+from builtins import object, range, str
```
Comment (mmphego) [inline nosekatreport/plugin.py]: Perhaps only import `object`.

File: nosekatreport/plugin.py [hunk 8]
```diff
@@ -16,6 +15,14 @@
...
+
+import colors
+import numpy as np
+# Copyright (c) 2017 National Research Foundation (South African Radio Astronomy Observatory)
```
Comment (mmphego) [inline nosekatreport/plugin.py]: Top of the file

File: nosekatreport/tests/test_aqf.py [hunk 9]
```diff
@@ -1,9 +1,16 @@
...
+from future import standard_library
+standard_library.install_aliases()
+
 import unittest
+from builtins import str
 
-from nosekatreport import (Aqf, aqf_vr, system, slow, intrusive,
-                           site_only, site_acceptance)
+# Copyright (c) 2017 National Research Foundation (South African Radio Astronomy Observatory)
+# BSD license - see LICENSE for details
```
Comment (mmphego) [inline nosekatreport/tests/test_aqf.py]: Ditto

File: nosekatreport/tests/test_aqf.py [hunk 10]
```diff
@@ -1,9 +1,16 @@
-# Copyright (c) 2017 National Research Foundation (South African Radio Astronomy Observatory)
-# BSD license - see LICENSE for details
+from __future__ import (absolute_import, division, print_function)
+
+from future import standard_library
+standard_library.install_aliases()
+
 import unittest
+from builtins import str
```
Comment (mmphego) [inline nosekatreport/tests/test_aqf.py]: Ditto

File: setup.py [hunk 11]
```diff
@@ -1,35 +1,53 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
```
Comment (mmphego) [inline setup.py]: Remove `unicode_literals`, 
https://python-future.org/unicode_literals.html#unicode-literals

File: setup.py [hunk 12]
```diff
@@ -1,35 +1,53 @@
...
+
 # Copyright (c) 2017 National Research Foundation (South African Radio Astronomy Observatory)
 # BSD license - see LICENSE for details
```
Comment (mmphego) [inline setup.py:2]: Top of the file

File: setup.py [hunk 13]
```diff
@@ -1,35 +1,53 @@
...
-    long_description=open('README.rst').read(),
-    author='cam',
-    author_email='cam@ska.ac.za',
-    license='BSD',
-    packages=find_packages(exclude=['ez_setup']),
-    install_requires=['Nose>=0.11.0', 'traceback2', 'ansicolors', 'numpy'],
-    url='',
+    name="nosekatreport",
+    description="Nose plugin for writing an annotated test report",
+    long_description=open("README.rst").read(),
```
Comment (mmphego) [inline setup.py]: Adopt this,
See: https://github.com/ska-sa/katportalclient/blob/master/setup.py#L9+L23

File: setup.py [hunk 14]
```diff
@@ -1,35 +1,53 @@
...
+    author="cam",
+    author_email="cam@ska.ac.za",
+    license="BSD",
+    packages=find_packages(exclude=["ez_setup"]),
+    url="",
```
Comment (mmphego) [inline setup.py]: Since this package is Open Source, add this URL: https://github.com/ska-sa/nosekatreport/

File: setup.py [hunk 15]
```diff
@@ -1,35 +1,53 @@
...
-        'Programming Language :: Python :: 2.7',
-        'Topic :: Software Development :: Testing',
-        'Topic :: Software Development :: Documentation'
+    classifiers=[
+        "Development Status :: 2 - Pre-Alpha",
+        "Environment :: Console",
+        "Intended Audience :: Developers",
+        "Natural Language :: English",
+        "License :: OSI Approved :: GNU General Public License (GPL)",
+        "Operating System :: OS Independent",
```
Comment (mmphego) [inline setup.py:38]: See: https://github.com/ska-sa/katportalclient/blob/master/setup.py#L38+L49

File: tox.ini [hunk 16]
```diff
@@ -0,0 +1,15 @@
+[tox]
```
Comment (mmphego) [inline tox.ini:1]: Update travis.yml

https://github.com/ska-sa/nosekatreport/commit/4a6d6d0ae545a49178bda3d1c12c78ca8e45f594

**Review by mmphego — CHANGES_REQUESTED** [review]

File: nosekatreport/plugin.py (hunk 6 above)
Comment (amakhaba) [reply nosekatreport/plugin.py:1]: Mistake i guess.

File: tox.ini (hunk 16 above)
Comment (amakhaba) [reply tox.ini:1]: update `travis.yml` ? or `tox.ini` ?

File: tox.ini [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+skip_missing_interpreters = true
+envlist = py{27, 36}
+
+indexserver =
+    default = http://pypi.camlab.kat.ac.za/pypi/trusty
```
Comment (mmphego) [inline tox.ini]: Remove `indexserver`, as we pulling all packages from `pypi`

File: setup.py [hunk 18]
```diff
@@ -1,35 +1,68 @@
...
+        "Topic :: Software Development :: Testing",
+        "Topic :: Software Development :: Documentation"
         ],
+    platforms=["OS Independent"],
+    keywords="meerkat kat ska",
+    python_requires=">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, <4",
+    setup_requires=["katversion"],
+    use_katversion=True,
+    install_requires=[
+        "elasticsearch>=5.4.0",
```
Comment (mmphego) [inline setup.py]: Why does `nosekatreport` need `elasticsearch` and `tornado`?

File: nosekatreport/tests/test_aqf.py [hunk 19]
```diff
@@ -1,9 +1,16 @@
+
```
Comment (mmphego) [inline nosekatreport/tests/test_aqf.py:1]: Why do you have an empty line?

File: nosekatreport/plugin.py (hunk 8 above)
Comment (mmphego) [reply nosekatreport/plugin.py]: Why is this marked resolved, but unfixed?

File: nosekatreport/plugin.py (hunk 6 above)
Comment (mmphego) [reply nosekatreport/plugin.py:1]: Why is this marked resolved, but unfixed?

File: .travis.yml [hunk 20]
```diff
@@ -1,7 +1,15 @@
...
+   -    "2.7"
+   -    "3.6"
+
+env:
+    KATPACKAGE=${PWD##*/}
+    test_flags="--with-xunit --with-xcoverage --cover-package=${KATPACKAGE} --cover-inclusive"
+ 
 install:
-  - pip install .
+  -     pip install tox-travis
```
Comment (mmphego) [inline .travis.yml]: Too many spaces, please use 4 spaces after the hyphen.

File: nosekatreport/__init__.py [hunk 21]
```diff
@@ -2,9 +2,17 @@
...
 
 """Root of the nosekatreport package."""
+from __future__ import (absolute_import, division, print_function)
+
+from future import standard_library
+standard_library.install_aliases()
 
 # BEGIN VERSION CHECK
 # Get package version when locally imported from repo or via -e develop install
+
```
Comment (mmphego) [inline nosekatreport/__init__.py:12]: Why did you move the imports?

See regarding imports [CAM Developer's Guide](https://docs.google.com/document/d/1aZoIyR9tz5rCWr2qJKuMTmKp2IzHlFjrCFrpDDHFypM/edit#heading=h.y0sdkqkmrbb3)

Comment (amakhaba) [reply nosekatreport/__init__.py:12]: This resulted from the futurize stage 0 application.

**Comment by tockards** [comment]
Once you addressed @mmphego  comment and we haved tested this by generating a report from AQF results. 
I will be happy to accept the PR.

File: nosekatreport/__init__.py (hunk 21 above)
Comment (mmphego) [reply nosekatreport/__init__.py:12]: Kindly fix this, please. Follow the guideline else refer below

We expand this to:
- standard library imports
- related third party imports
- ‘kat’ imports, i.e. packages that are in SVN
- **local application/library specific imports**	

> "The first rule of any technology used in a business is that automation applied to an efficient operation will magnify the efficiency. The second is that automation applied to an inefficient operation will magnify the inefficiency." - Bill Gates

**Note:** Automate and make manual fixes were needed.

**Review by mmphego — APPROVED** [review]
LGTM, Before you merge let's run AQF with report generation.

File: nosekatreport/__init__.py (hunk 21 above)
Comment (amakhaba) [reply nosekatreport/__init__.py:12]: Latest Commit addresses this.

**Comment by mmphego** [comment]
It would be nice if we could add `coverage` badge on the readme file.

See: https://docs.travis-ci.com/user/coveralls/

**Comment by mmphego** [comment]
Please `Squash and merge` after the report generation is done!

**Review by tockards — APPROVED** [review]
I tested the report generation on the aqf node by changing the branch of nosekatreport to this branch on and running report generation again.
It was successful in the creation of the reports.
